DM-50727: Change pruneDatasets to only prune the requested datasets

timj opened the pull request:
Rather than emptyTrash always trying to clear the entire trash table, add the ability to constrain the removal by a specific set of refs (that generally are the refs that were just pruned). Also take the opportunity to chunk deletes into 50,000 files to reduce the size of transactions that could involve a million files.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

timj committed:
Change pruneDatasets to only prune the requested datasets

Rather than emptyTrash always trying to clear the entire trash
table, add the ability to constrain the removal by a specific
set of refs (that generally are the refs that were just pruned).
Also take the opportunity to chunk deletes into 50,000 files
to reduce the size of transactions that could involve a million
files.

timj committed:
Change emptyTrash to return list of removed URIs

Also reorganize error handling to make it clearer and remove
try/except blocks.

timj committed:
Add dry_run flag to emptyTrash

timj committed:
Simplify and unify delete logic in emptyTrash

Use URIs directly rather than Location to avoid requesting a
Zip file be deleted multiple times due to different fragment.

timj committed:
If only deleting one artifact in trash emptying report that artifact

timj committed:
Add test for emptyTrash parameters

andy-slac commented on python/lsst/daf/butler/registry/bridge/monolithic.py:
```diff
@@ -249,6 +252,10 @@ def join_records(
 
+        # Only keep rows that match the external override selection.
+        if selected_ids is not None:
+            rows = [row for row in rows if row["dataset_id"] in selected_ids]
```
I am worried that this still leaves a chance for a deadlock in case one uses `admin empty-trash` concurrently with regular removal. We probably need to use explicit locking to avoid that. What I would do is something like (line 248 above):

    info_in_trash = join_records(records_table._table.select(), self._tables.dataset_location_trash)
    if selected_ids:
        info_in_trash = info_in_trash.where(self._tables.dataset_location_trash.columns["dataset_id"].in_(selected_ids))
    info_in_trash = info_in_trash.with_for_update(skip_locked=True)

And I think that we want similar `.where()` for `items_in_trash` below too.

andy-slac reviewed (approved):
Looks good, though I suspect it may still deadlock under rare circumstances. I left one suggestion if you want to try it.

andy-slac committed:
Try to avoid deadlocks by limiting the scope of SELECT.

When list of dataset IDs is passed to `emptyTrash` it now runs SELECT
using only those IDs, this avoids taking locks on all records in trash
table. Additionally FOR UPDATE is added to that SELECT so that there
is no need to upgrade locks later.

timj commented on python/lsst/daf/butler/registry/bridge/monolithic.py:
```diff
@@ -243,6 +246,11 @@ def join_records(
+            info_in_trash = info_in_trash.where(
+                self._tables.dataset_location_trash.columns["dataset_id"].in_(selected_ids)
+            )
+        info_in_trash = info_in_trash.with_for_update(skip_locked=True)
```
@dhirving are you thinking that there should also be an order_by clause here to sort the results (eg by path if sorting by UUID is less efficient)?  cc/ @andy-slac

dhirving replied:
I don't know enough about the internals of how Postgres takes row locks.  But I doubt an `ORDER BY` will give us any extra guarantees here, and it will probably hurt performance.

timj committed:
Rearrange child removal loop to try to avoid deadlocks

Since the order previously was not deterministic there was a chance
that two simultaneous calls to butler remove-runs could deadlock.

timj committed:
Use different variable name to appease mypy

timj committed:
Add news fragment
